src/module_event_handlers.c

```diff
@@ -0,0 +1,244 @@
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "module_event_handlers.h"
+#include <pthread.h>
+#include <assert.h>
+#include <stdbool.h>
+#include "graph/graphcontext.h"
+#include "graph/graphmetacontext.h"
+#include "serializers/graphcontext_type.h"
+#include "serializers/graphmeta_type.h"
+#include "util/uuid.h"
+
+extern GraphContext **graphs_in_keyspace;  // Global array tracking all extant GraphContexts.
+extern bool
+process_is_child;              // Flag indicating whether the running process is a child.
+extern uint64_t entities_threshold;        // The limit of number of entities encoded at once.
+extern uint redis_major_version;           // The redis server major version.
+
+/* This callback invokes once rename for a graph is done. Since the key value is a graph context
+ * which saves the name of the graph for later key accesses, this data must be consistent with the key name,
+ * otherwise, the graph context will remain with the previous graph name, and a key access to this name might
+ * yield an empty key or wrong value. This method changes the graph name value at the graph context to be
+ * consistent with the key name. */
+static int _RenameGraphHandler(RedisModuleCtx *ctx, int type, const char *event,
+							   RedisModuleString *key_name) {
+	if(type != REDISMODULE_NOTIFY_GENERIC) return REDISMODULE_OK;
+	if(strcasecmp(event, "RENAME_TO") == 0) {
+		RedisModuleKey *key = RedisModule_OpenKey(ctx, key_name, REDISMODULE_WRITE);
+		if(RedisModule_ModuleTypeGetType(key) == GraphContextRedisModuleType) {
+			GraphContext *gc = RedisModule_ModuleTypeGetValue(key);
+			size_t len;
+			const char *new_name = RedisModule_StringPtrLen(key_name, &len);
+			GraphContext_Rename(gc, new_name);
+		}
+		RedisModule_CloseKey(key);
+	}
+	return REDISMODULE_OK;
+}
+
+//------------------------------------------------------------------------------
+// Meta keys API
+//------------------------------------------------------------------------------
+
+// Checks if the graph name contains a tag between curly braces.
+static bool _GraphContext_NameContainsTag(const GraphContext *gc) {
+	const char *left_curly_brace = strstr(gc->graph_name, "{");
+	if(left_curly_brace) {
+		const char *right_curly_brace = strstr(left_curly_brace, "}");
+		if(right_curly_brace) {
+			return true;
+		}
+	}
+	return false;
+}
+
+// Calculate how many virtual keys are needed to represent the graph.
+static uint64_t _GraphContext_RequiredMetaKeys(const GraphContext *gc) {
+	uint64_t required_keys = 0;
+	required_keys += ceil((double)Graph_NodeCount(gc->g) / entities_threshold);
+	required_keys += ceil((double)Graph_EdgeCount(gc->g) / entities_threshold);
+	required_keys += ceil((double)Graph_DeletedNodeCount(gc->g) / entities_threshold);
+	required_keys += ceil((double)Graph_DeletedEdgeCount(gc->g) / entities_threshold);
+	return required_keys;
+}
+
+static void _CreateGraphMetaKeys(RedisModuleCtx *ctx, GraphContext *gc) {
```

Comment:
This whole function should contain a lot more explanatory comments.

I feel like this approach is overcomplicated, however. We want to create N keys here, but don't actually care about their names except insofar as we want to delete them later. Why not just name them as `{i}graph_name`?

Reply:
done